File: wsb_signals/assets.py
"""Build the ticker whitelist from Alpaca's asset universe (Phase 0.4).

The extractor validates bare candidate tokens against this set, so we keep only
*listed, tradable* US equities/ETFs and drop OTC pink-sheets (they add ~1k
regex-matching junk symbols). A $-cashtag bypasses the whitelist, and the stoplist
still overrides common-word tickers (ALL, DD, IT). Refresh weekly (architecture §2.2).
"""
from __future__ import annotations

from pathlib import Path

import httpx

from .log import get_logger

log = get_logger("wsb.assets")


def fetch_alpaca_assets(
    key: str,
    secret: str,
    *,
    endpoint_url: str,
    user_agent: str = "wsb-signals/0.0.1",
    include_otc: bool = False,
    timeout: int = 60,
) -> list[tuple[str, str]]:
    """Active, tradable us_equity (symbol, name) pairs from /v2/assets (sorted, deduped by symbol)."""
    resp = httpx.get(
        endpoint_url.rstrip("/") + "/assets",
        params={"status": "active", "asset_class": "us_equity"},
        headers={"APCA-API-KEY-ID": key, "APCA-API-SECRET-KEY": secret, "User-Agent": user_agent},
        timeout=timeout,
    )
    resp.raise_for_status()
    by_sym: dict[str, str] = {}
    for a in resp.json():
        if not a.get("tradable"):
            continue
        if not include_otc and a.get("exchange") == "OTC":
            continue
        sym = (a.get("symbol") or "").strip()
        if sym and sym not in by_sym:
            by_sym[sym] = (a.get("name") or "").strip()
    return sorted(by_sym.items())


def build_whitelist(
    key: str,
    secret: str,
    *,
    endpoint_url: str,
    out_path: Path,
    user_agent: str = "wsb-signals/0.0.1",
    include_otc: bool = False,
) -> list[tuple[str, str]]:
    """Fetch the universe, write symbols to `out_path`, and return the (symbol, name) pairs.

    The caller persists the names to the `ticker_names` DuckDB table (displayed everywhere).
    """
    assets = fetch_alpaca_assets(
        key, secret, endpoint_url=endpoint_url, user_agent=user_agent, include_otc=include_otc
    )
    out_path.parent.mkdir(parents=True, exist_ok=True)
    header = (
        "# Ticker whitelist — Alpaca active, tradable us_equity"
        + ("" if include_otc else " (non-OTC)")
        + ".\n"
        "# DERIVED + refreshable, gitignored. Regenerate weekly: `wsb build-whitelist`.\n"
        "# One symbol per line; company names live in the ticker_names DuckDB table. See architecture §2.2.\n"
    )
    out_path.write_text(header + "\n".join(s for s, _ in assets) + "\n")
    return assets

File: packages/worker/src/assets.ts
/**
 * Ticker whitelist builder — TS port of the frozen v0.0.1 `wsb_signals/assets.py`
 * (porting-spec §4 extension, M5). Fetches the Alpaca active us_equity universe,
 * dedupes by symbol (first occurrence wins), and writes `symbols.txt` for the extractor.
 *
 * Parity notes:
 *  - `tradable:false` rows are dropped (Python `if not a.get("tradable"): continue`).
 *  - OTC rows are dropped by default (`!includeOtc && a.exchange === 'OTC'`).
 *  - Deduplication is first-occurrence-wins (Python `if sym not in by_sym`).
 *  - Output is sorted ascending by symbol (code-point order — Python `sorted(by_sym.items())`).
 *  - Non-200 throws (Python `resp.raise_for_status()`).
 *  - Manual AbortController timeout (same pattern as ingest.ts `get`).
 */
import { mkdirSync, writeFileSync } from 'node:fs'
import { dirname } from 'node:path'

import { fetch } from 'undici'

import { log } from './logger'

export async function fetchAlpacaAssets(
  key: string,
  secret: string,
  opts: {
    endpointUrl: string
    userAgent?: string
    includeOtc?: boolean
    timeoutMs?: number
  },
): Promise<Array<[string, string]>> {
  const {
    endpointUrl,
    userAgent = 'wsb-signals/0.0.1',
    includeOtc = false,
    timeoutMs = 60_000,
  } = opts

  const base = endpointUrl.replace(/\/+$/, '')
  const url = `${base}/assets?status=active&asset_class=us_equity`

  const ac = new AbortController()
  const timer = setTimeout(() => ac.abort(new Error(`timeout after ${timeoutMs}ms`)), timeoutMs)
  timer.unref?.()

  let res: Awaited<ReturnType<typeof fetch>>
  try {
    res = await fetch(url, {
      method: 'GET',
      headers: {
        'APCA-API-KEY-ID': key,
        'APCA-API-SECRET-KEY': secret,
        'User-Agent': userAgent,
      },
      signal: ac.signal,
    })
  } finally {
    clearTimeout(timer)
  }

  if (res.status !== 200) {
    const body = await res.text().catch(() => '')
    throw new Error(`Alpaca /assets returned HTTP ${res.status}: ${body.slice(0, 200)}`)
  }

  const data = (await res.json()) as Array<Record<string, unknown>>

  const bySym: Map<string, string> = new Map()
  for (const a of data) {
    if (!a['tradable']) continue
    if (!includeOtc && a['exchange'] === 'OTC') continue
    const sym = String(a['symbol'] ?? '').trim()
    if (sym && !bySym.has(sym)) {
      bySym.set(sym, String(a['name'] ?? '').trim())
    }
  }

  // Sort ascending by symbol (code-point order — Python `sorted(by_sym.items())`)
  const pairs: Array<[string, string]> = [...bySym.entries()]
  pairs.sort((a, b) => (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0))
  return pairs
}

export async function buildWhitelist(
  key: string,
  secret: string,
  opts: {
    endpointUrl: string
    outPath: string
    userAgent?: string
    includeOtc?: boolean
  },
): Promise<Array<[string, string]>> {
  const { endpointUrl, outPath, userAgent, includeOtc = false } = opts

  log.info({ endpointUrl }, 'fetching Alpaca asset universe…')
  const assets = await fetchAlpacaAssets(key, secret, { endpointUrl, userAgent, includeOtc })

  mkdirSync(dirname(outPath), { recursive: true })

  const suffix = includeOtc ? '' : ' (non-OTC)'
  const header =
    `# Ticker whitelist — Alpaca active, tradable us_equity${suffix}.\n` +
    '# DERIVED + refreshable, gitignored. Regenerate weekly: `pnpm -C packages/worker build-whitelist`.\n' +
    '# One symbol per line; company names live in the ticker_names table. See architecture §2.2.\n'

  const symbols = assets.map(([sym]) => sym)
  writeFileSync(outPath, header + symbols.join('\n') + '\n', 'utf8')

  return assets
}
